Cookbook/Proposals/Danger-Rule-Xcodeproj.md

```diff
@@ -0,0 +1,132 @@
+# Danger Rules for pbxproj
+
+* Author: Olivier Halligon
+* Review Manager: TBD
+
+## Introduction
+
+This proposal suggests to add some Danger rules checking for `xcodeproj` consistenty.
+
+It is a follow-up on [#110 - Integrate Danger](https://github.com/Babylonpartners/ios-playbook/pull/110).
+
+## Motivation
+
+`pbxproj` project format is hard to parse, which means it's easy to miss some mistakes that could happen when an Xcode project has changed during a PR.
+
+This means that some mistakes have ended up in `develop` in the past, like:
+
+* `(null)` references and `"Recovered References"` groups in `pbxproj` files
+* Invalid or Orphan UUIDs in `pbxproj` files (†)
+* Files being added as part of a target membership while they shouldn't:
+  * Snapshot files
+  * xcconfig files
+  * Info.plist files
+* Files expected to be in a target but are not (especially tests that we might forget to include and are thus not run)
+
+(†) For example this is the kind of warnings we can see when running `bundle exec pod install` and the projects manipulated by CocoaPods contain some invalid UUIDs:
+
+```
+`<PBXGroup path=`Tests` UUID=`F26256AC222EC21500DD5E7B`>` attempted to initialize an object with an unknown UUID. `5FFC934E2296E0980088BFAA` for attribute: `children`. This can be the result of a merge and  the unknown UUID is being discarded.
+`<PBXGroup path=`DesignLibrary` UUID=`F26256A8222EC21500DD5E7B`>` attempted to initialize an object with an unknown UUID. `5F76D723229456BA00EC6CBC` for attribute: `children`. This can be the result of a merge and  the unknown UUID is being discarded.
+`<PBXSourcesBuildPhase UUID=`A65543DE1F50510600BAABE4`>` attempted to initialize an object with an unknown UUID. `5F76D724229456BA00EC6CBC` for attribute: `files`. This can be the result of a merge and  the unknown UUID is being discarded.
+`<PBXSourcesBuildPhase UUID=`58E92B48208799C90037BD6A`>` attempted to initialize an object with an unknown UUID. `5FFC934F2296E0980088BFAA` for attribute: `files`. This can be the result of a merge and  the unknown UUID is being discarded.
+```
+
+Some of those mistakes, especially `(null)`, "Recovered References" and invalid UUIDs in `pbxproj`, are generally the result of merge conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.
+
+Other mistakes like unexpected files in target membership could lead to files ending up in the final bundle uploaded to the AppStore, and in addition to being useless in the final `ipa`, bloat the app size for no reason.
+
+## Proposed solution
+
+We propose to use Danger rules to analyze the pbxproj and find all those issues during review:
```

Comment:
Make sure that the Danger comment contains suggestions about how to solve the issues when they happen.